ethereum/contracts/TablelandRigs.sol

```diff
@@ -354,6 +390,231 @@ contract TablelandRigs is
+            ) {
+                parkRig(i);
+            }
+        }
+        // Assign a new pilot to the Rig, incrementing the previous pilot's `index` by `1`
+        _pilots[tokenId] = RigPilot(
+            _pilots[tokenId].index + 1,
+            uint64(block.number),
+            pilotContract,
+            pilotTokenId
+        );
+        _rigsStatus[tokenId - 1] = 1;
+        // Insert the pilot into the Tableland `rig_pilot_sessions` table
+        // TODO Remove the following, but for context, the statement
+        /**
+            INSERT INTO rigs_pilots (
+                rig_id, owner, pilot_contract, 
+                pilot_id, start_time
+            ) 
+            VALUES 
+            (
+                tokenId, ownerOf(tokenId), pilotContract,
+                pilotTokenId, block.number
+            )
+         */
+        TablelandDeployments.get().runSQL(
+            address(this),
+            _rigPilotSessionsTableId,
+            SQLHelpers.toInsert(
+                RIG_PILOT_SESSIONS_PREFIX,
+                _rigPilotSessionsTableId,
+                "rig_id,owner,pilot_contract,pilot_id,start_time",
+                string.concat(
+                    StringsUpgradeable.toString(uint64(tokenId)),
+                    ",",
+                    StringsUpgradeable.toHexString(ownerOf(tokenId)),
+                    ",",
+                    Strings.toHexString(pilotContract),
+                    ",",
+                    StringsUpgradeable.toString(uint64(pilotTokenId)),
+                    ",",
+                    StringsUpgradeable.toString(uint64(block.number))
+                )
+            )
+        );
+        emit Piloted(tokenId);
+    }
+
+    /**
+     * @dev See {ITablelandRigs-parkRig}.
+     */
+    function parkRig(uint256 tokenId) public {
+        // Check the Rig `tokenId` exists.
+        if (!_exists(tokenId)) revert OwnerQueryForNonexistentToken();
+        // Verify `msg.sender` is authorized to park the specified Rig
+        if (ownerOf(tokenId) != _msgSenderERC721A()) {
+            if (address(this) != _msgSenderERC721A()) {
+                revert InvalidRigOwnership(tokenId);
+            }
+        }
+        // Ensure Rig is currently being piloted
+        uint64 startTime = _pilots[tokenId].startTime;
+        if (startTime == 0) revert RigIsParked(tokenId);
+        // Ensure the Rig has trained; `index` of `1` indicates training is occurring
+        // If the Rig is training, check if 30 days has elapsed (the required training time)
+        (bool isValidSubtraction, uint256 blockDifference) = SafeMathUpgradeable
+            .trySub(block.number, 172800);
+        if (
+            !isValidSubtraction ||
+            _pilots[tokenId].index == 0 ||
+            (_pilots[tokenId].index == 1 &&
+                !(blockDifference > _pilots[tokenId].startTime))
+        ) {
+            revert RigIsNotTrained(tokenId);
+        }
```

Comment:
shouldn't we let rigs park even if training isn't finished? they could just start from 0 again

Reply:
@sanderpick yeah, that could work too...it basically ends up with the same result, minus the locking aspect, which makes it flexible for the user. we'd just need to make sure the frontend makes a callout to _not_ park the Rig if 30 consecutive days of training hasn't happened yet, otherwise, they'll start over (cc @datadanne).

Reply:
one random idea for the extra data in the `Pilot` struct. we're currently tracking the `Pilot.index` of `1` to check if training has happened or not for a single Rig, and training only happens once. instead, we could maybe make training a requirement for every new Rig owner? namely, have simple `isTrained` field in the struct as a `uint8`, and on `_afterTokenTransfer`, this gets reset to `0`. idea being, the training isn't a one time thing, which makes it feel "nicer" as part of the code. just an idea.

(not sure if i originally misinterpreted the design, but this feels more natural too since a pilot should have to train to use a Rig)

Reply:
see https://github.com/tablelandnetwork/rigs/pull/233#discussion_r996351264. it think we can handle this in the way described there. i like the idea of "training the rig" vs. pilot (they are sentient machines :)), which adds some friction to getting flight time for people that have tons of rigs, vs. them being able to train on one rig for piloting many.